users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py

```diff
@@ -58,60 +54,51 @@ def __init__(
         if preemphasis:
             self._perturbations.append(functools.partial(self.preemphasis, factor=PerturbationFactor(**preemphasis)))
-        if codecs:
-            self._perturbations.append(functools.partial(self.apply_codecs, codecs=codecs))
 
     def run(self, audio, sample_rate, random_state):
-        audio = torch.from_numpy(audio).unsqueeze(0).to(torch.float32)
-        for perturbation in self._perturbations:
-            audio = perturbation(audio, sample_rate, random_state)
-        audio = audio.numpy().squeeze()
+        import numpy as np
         assert isinstance(audio, np.ndarray)
         assert len(audio.shape) == 1
+        audio = audio.astype(np.float32)
+        audio = self.sox(audio, sample_rate, random_state)
         return audio
 
-    def sox(self, audio, sample_rate, random_state, sox_effects):
-        sox_effects = sox_effects or []
+    def sox(self, audio, sample_rate, random_state):
+        import random
+        import sox
         speed = False
-        if self._speed is not None:
-            if random_state.random() < self._speed.prob:
-                factor = random_state.random() * (self._speed.max - self._speed.min) + self._speed.min
-                sox_effects.append(["speed", str(factor)])
-                sox_effects.append(["rate", str(sample_rate)])
-                speed = True
-        if self._tempo is not None:
-            if random_state.random() < self._tempo.prob and not speed:
-                factor = random_state.random() * (self._tempo.max - self._tempo.min) + self._tempo.min
-                sox_effects.append(["tempo", str(factor)])
-        audio, _ = torchaudio.sox_effects.apply_effects_tensor(audio, sample_rate, sox_effects)
+        tfm = sox.Transformer()
+        if self._speed is not None and random.random() < self._speed.prob:
+            factor = random.random() * (self._speed.max - self._speed.min) + self._speed.min
+            tfm.speed(factor)
+            speed = True
+        if self._tempo is not None and random.random() < self._tempo.prob and not speed:
+            factor = random.random() * (self._tempo.max - self._tempo.min) + self._tempo.min
+            #tfm.tempo(factor) is more efficiant for larger factors(>1.1).
+            tfm.stretch(factor)
+        for effect in self._sox_effects:
+            effect_name, *params = effect
+            getattr(tfm, effect_name)(*params)
+        audio = tfm.build_array(input_array=audio, sample_rate_in=sample_rate)
         return audio
 
     @staticmethod
     def preemphasis(audio, sample_rate, random_state, factor):
+        import numpy as np
         if random_state.random() < factor.prob:
             preemphasis_coefficient = random_state.random() * (factor.max - factor.min) + factor.min
-            audio = torchaudio.functional.preemphasis(audio, coeff=preemphasis_coefficient)
+            return np.append(audio[0], audio[1:] - preemphasis_coefficient * audio[:-1])
```

Comment:
Is there no dedicated function for this in sox?

Reply:
No there is not. Only HighPass which is slightly different.

Reply:
Did you verify that this gives the same result as before? I'm not sure if it is common to include `audio[0]` like that.

Reply:
Changed it so it is identical to the torch implementation